working/0698 - Enhanced Default Constructors/proposal.md

```diff
@@ -0,0 +1,166 @@
+- For each named parameter of *s*, *c* has a corresponding parameter with the same name, type, optionality and default value (if optional), unless a named initializing formal parameter with the same name was declared above (in which case the parameter of *s* must be optional per requirement 4. above).
+- For each positional parameter of *s*, whether optional or not, $c$ has a corresponding positional parameter with the same position, type, optionality and default value (if optional). For documentation purposes, the parameter will have the same name as the corresponding positional parameter of *s* unless that conflicts with one of the named initializing formals added in the previous point. In that case, the name will be modified by adding a `$` and a positive decimal integer numeral with no leading zeros after the name, where the numeral represents the smallest positive (&ge; 1) integer sufficient to make the name not be the same as any of the named parameters of the constructor, not the same as the name of any later positional parameter of *s*, and not the same as any earlier positional parameter of *c*.
+
+and with an initializer list consisting of a single `super` invocation of either `super`, if *s* is named `S`, or `super.id`, if *s* is named `S.id`, with one argument for each parameter of *c* corresponding to a parameter of *s*, forwarding the value of the *c* parameter as an argument to the corresponding *s* parameter.
+
+The constructor *c* is `const` if *s* is `const` and *C* does not declare *any* instance variables.
+
+It is a *compile-time error* if there are no superclass constructors satisfying the requirements. (This is not a necessity, it's completely valid for a class to have no generative constructors, which is what currently happens if a class declares only factory constructors. We require the author to explicitly opt in to this behavior because *if* it is unintentional, it is a very hard-to-spot error.)
+
+It is a *compile-time error* to derive a mixin from a class unless its superclass is `Object`, it declares no generative constructors, and it declares no potentially non-nullable instance members without an initializer expression. (For NNBD-legacy code, it must declare no final fields without initializer expressions.)
+
+### Examples
+
+The following class declarations will be valid:
+
+```dart
+class Person {
+  final String firstName;
+  final String lastName;
+  final int age;
+}
+
+class Fireman extends Person {
+  final int yearsOfService;
+}
+```
+
+You can instantiate a `Fireman` as:
+
+```dart
+ Fireman(firstName: "John", lastName: "Doe", age: 37, yearsOfService: 12)
+```
+
+This is still verbose, because of the named arguments.
```

Comment:
I don't think that's particularly verbose. :) Flutter users will probably feel this looks entirely normal.